linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -673,19 +678,54 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
   private def transformClosureCommon(arrow: Boolean,
       captureParams: List[ParamDef], params: List[ParamDef],
       restParam: Option[ParamDef], body: Tree,
-      newCaptureValues: List[Tree])(
-      implicit scope: Scope, pos: Position): Closure = {
+      tcaptureValues: List[PreTransform])(cont: PreTransCont)(
+      implicit scope: Scope, pos: Position): TailRec[Tree] = {
 
     val thisType = if (arrow) NoType else AnyType
-    val (allNewParams, newBody) = transformIsolatedBody(None, thisType,
-        captureParams ++ params ++ restParam, AnyType, body, scope.implsBeingInlined)
-    val (newCaptureParams, newParams, newRestParam) = {
-      val (c, t) = allNewParams.splitAt(captureParams.size)
-      if (restParam.isDefined) (c, t.init, Some(t.last))
-      else (c, t, None)
+
+    val (allNewParamsWithUsage, newBody) =
+      transformIsolatedBody(None, thisType, captureParams ++ params ++ restParam, AnyType, body,
+        scope.implsBeingInlined)
+
+    val captureParamCount = captureParams.size
+
+    val (newParams, newRestParam) = {
+      val t = allNewParamsWithUsage.iterator.drop(captureParamCount).map(_._1).toList
+      if (restParam.isDefined) (t.init, Some(t.last))
+      else (t, None)
+    }
+
+    val (newCaptureParams, newCaptureValues, unusedCaptureValues) = {
+      val newCaptureParams = List.newBuilder[ParamDef]
+      val newCaptureValues = List.newBuilder[Tree]
+      val unusedCaptureValues = List.newBuilder[Binding]
+      var unused = 0
+
+      allNewParamsWithUsage
+        .iterator
+        .take(captureParamCount)
```

Comment:
It took me a while to understand that this `take`, and the `drop` a few lines above, are related to the combination of param defs with `++` that we do in the beginning. So IIUC they replaced the `splitAt`, but in two separate places. Consider restoring a single `splitAt` call, close to the `++`'es that they cancel out.